### PR TITLE
Rholang: Normalizer: Add Name normalization

### PR DESCRIPTION
**Pull request by guardbotmk3**
Mostly, name normalization delegates to process normalization except for vars
and wildcards. Add tests for var, wildcard, and a sampling of delegation tests.

### CODE REVIEW

File: rholang/src/test/scala/rholang/interpretr/NormalizeTest.scala [hunk 1]
```diff
@@ -51,7 +51,7 @@ class NilVisitSpec extends FlatSpec with Matchers {
   }
```
Comment (KentShikama) [inline rholang/src/test/scala/rholang/interpretr/NormalizeTest.scala]: The folder name has a typo.

File: rholang/src/main/scala/rholang/interpreter/Normalize.scala [hunk 2]
```diff
@@ -22,18 +22,58 @@ trait GroundNormalizeVisitor extends AbsynGround.Visitor[Ground, Any]
   override def visit( gi: GroundUri, n: Any ) : Ground = GUri(gi.uri_)
 }
 
+trait NameNormalizeVisitor extends Name.Visitor[NameVisitOutputs, NameVisitInputs] {
+  val procVisitor : Proc.Visitor[ProcVisitOutputs, ProcVisitInputs]
```
Comment (henrytill) [inline rholang/src/main/scala/rholang/interpreter/Normalize.scala]: It is customary to place the colon for a type annotation directly after the identifier name (or parentheses, in the case of a function definition).  Should be:
```scala
val procVisitor: Proc.Visitor[ProcVisitOutputs, ProcVisitInputs]
```

File: rholang/src/main/scala/rholang/interpreter/Normalize.scala [hunk 3]
```diff
@@ -22,18 +22,58 @@ trait GroundNormalizeVisitor extends AbsynGround.Visitor[Ground, Any]
...
+              ChanVar(FreeVar(newBindingsPair._2(0))),
+              newBindingsPair._1)
+          case _ => throw new Error(
+            "Free variable used as binder may not be used twice.")
+        }
+      }
+    }
+  }
+
+  override def visit( n: NameQuote, input: NameVisitInputs ) = {
```
Comment (henrytill) [inline rholang/src/main/scala/rholang/interpreter/Normalize.scala]: Formatting.

File: rholang/src/main/scala/rholang/interpreter/Normalize.scala [hunk 4]
```diff
@@ -22,18 +22,58 @@ trait GroundNormalizeVisitor extends AbsynGround.Visitor[Ground, Any]
...
+      procVisitResult.knownFree)
+  }
+}
+
 trait ProcNormalizeVisitor
     extends Proc.Visitor[ProcVisitOutputs, ProcVisitInputs]
     with GroundNormalizeVisitor {
 
-  override def visit(p: PGround, input: ProcVisitInputs)
+  override def visit( p: PGround, input: ProcVisitInputs )
```
Comment (henrytill) [inline rholang/src/main/scala/rholang/interpreter/Normalize.scala]: Formatting.

File: rholang/src/main/scala/rholang/interpreter/Normalize.scala [hunk 5]
```diff
@@ -22,18 +22,58 @@ trait GroundNormalizeVisitor extends AbsynGround.Visitor[Ground, Any]
...
       : ProcVisitOutputs = {
     ProcVisitOutputs(
       input.par.copy(expr = p.ground_.accept(this, input) :: input.par.expr),
       input.knownFree)
   }
 
-  override def visit(p: PVar, input: ProcVisitInputs)
+  override def visit( p: PVar, input: ProcVisitInputs )
```
Comment (henrytill) [inline rholang/src/main/scala/rholang/interpreter/Normalize.scala]: Formatting.

File: rholang/src/main/scala/rholang/interpreter/Normalize.scala [hunk 6]
```diff
@@ -61,55 +101,72 @@ trait ProcNormalizeVisitor
     }
   }
 
-  override def visit(p: PNil, input: ProcVisitInputs) : ProcVisitOutputs = {
+  override def visit( p: PNil, input: ProcVisitInputs ) : ProcVisitOutputs = {
```
Comment (henrytill) [inline rholang/src/main/scala/rholang/interpreter/Normalize.scala]: Formatting.

File: rholang/src/main/scala/rholang/interpreter/Normalize.scala [hunk 7]
```diff
@@ -61,55 +101,72 @@ trait ProcNormalizeVisitor
...
     ProcVisitOutputs(input.par, input.knownFree)
   }
+
+  override def visit( p: PPar, input: ProcVisitInputs ) : ProcVisitOutputs = {
```
Comment (henrytill) [inline rholang/src/main/scala/rholang/interpreter/Normalize.scala]: Formatting.

File: rholang/src/main/scala/rholang/interpreter/Normalize.scala [hunk 8]
```diff
@@ -61,55 +101,72 @@ trait ProcNormalizeVisitor
...
+  override def visit( p: PAnd, input: ProcVisitInputs ) : ProcVisitOutputs = ???
+  override def visit( p: POr, input: ProcVisitInputs ) : ProcVisitOutputs = ???
+  override def visit( p: PSend, input: ProcVisitInputs ) : ProcVisitOutputs = ???
+  override def visit( p: PContr, input: ProcVisitInputs ) : ProcVisitOutputs = ???
+  override def visit( p: PInput, input: ProcVisitInputs ) : ProcVisitOutputs = ???
+  override def visit( p: PChoice, input: ProcVisitInputs ) : ProcVisitOutputs = ???
+  override def visit( p: PMatch, input: ProcVisitInputs ) : ProcVisitOutputs = ???
+  override def visit( p: PIf, input: ProcVisitInputs ) : ProcVisitOutputs = ???
+  override def visit( p: PIfElse, input: ProcVisitInputs ) : ProcVisitOutputs = ???
+  override def visit( p: PNew, input: ProcVisitInputs ) : ProcVisitOutputs = ???
```
Comment (henrytill) [inline rholang/src/main/scala/rholang/interpreter/Normalize.scala]: Formatting 117-142.  

Should be like:
```scala
override def visit(p: PCollect, input: ProcVisitInputs): ProcVisitOutputs = ???
```

File: rholang/src/main/scala/rholang/interpreter/Normalize.scala [hunk 9]
```diff
@@ -22,18 +22,58 @@ trait GroundNormalizeVisitor extends AbsynGround.Visitor[Ground, Any]
   override def visit( gi: GroundUri, n: Any ) : Ground = GUri(gi.uri_)
 }
 
+trait NameNormalizeVisitor extends Name.Visitor[NameVisitOutputs, NameVisitInputs] {
```
Comment (henrytill) [inline rholang/src/main/scala/rholang/interpreter/Normalize.scala]: It's generally customary to put classes or traits in their own separate file.

See:
https://docs.scala-lang.org/style/files.html

If you feel like this falls under the "Multi-Unit File" provision, then you should probably use a lower camel-case name, as the article suggests.

File: rholang/src/test/scala/rholang/interpreter/NormalizeTest.scala [hunk 10]
```diff
@@ -0,0 +1,231 @@
+import coop.rchain.interpreter._
```
Comment (henrytill) [inline rholang/src/test/scala/rholang/interpreter/NormalizeTest.scala]: This file is missing a package declaration:
```scala
package coop.rchain.rholang.interpreter
```

(I understand that that other files use the package `coop.rchain.interpreter`, but those should probably be changed as well)

The path to this file should mirror the package name, just like in Java:
```
./src/test/scala/coop/rchain/rholang/interpreter/NormalizeTest.scala
```

File: rholang/src/main/scala/rholang/interpreter/Normalize.scala [hunk 11]
```diff
@@ -139,3 +210,12 @@ case class ProcVisitInputs(
   knownFree: DebruijnLevelMap[VarSort] )
 // Returns the update Par and an updated map of free variables.
 case class ProcVisitOutputs( par: Par, knownFree: DebruijnLevelMap[VarSort] )
+
+sealed trait ChanPosition
```
Comment (henrytill) [inline rholang/src/main/scala/rholang/interpreter/Normalize.scala]: See above comment about separate files.

File: rholang/src/main/scala/rholang/interpreter/Normalize.scala [hunk 12]
```diff
@@ -139,3 +210,12 @@ case class ProcVisitInputs(
...
+case object BindingPosition extends ChanPosition
+case object UsePosition extends ChanPosition
+
+case class NameVisitInputs(
+  env: DebruijnLevelMap[VarSort],
+  knownFree: DebruijnLevelMap[VarSort] )
+case class NameVisitOutputs( chan: Channel, knownFree: DebruijnLevelMap[VarSort] )
```
Comment (henrytill) [inline rholang/src/main/scala/rholang/interpreter/Normalize.scala]: Separate files?  Top-level object?

File: rholang/src/test/scala/rholang/interpreter/NormalizeTest.scala [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+import coop.rchain.syntax.rholang_mercury.Absyn.{Ground => AbsynGround, _}
+import org.scalatest._
+
+class BoolVisitorSpec extends FlatSpec with Matchers {
```
Comment (henrytill) [inline rholang/src/test/scala/rholang/interpreter/NormalizeTest.scala]: How about putting all the tests in a single class like:

```scala
class NormalizeTest extends FlatSpec with Matchers {

  val boolNormalizeVisitor: BoolNormalizeVisitor = new BoolNormalizeVisitor() {}

  "A BoolNormalizeVisitor" should "convert a BoolTrue to GBool(true)" in {
    new BoolTrue().accept(boolNormalizeVisitor, this) should be(GBool(true))
  }

  it should "convert a BoolFalse to GBool(false)" in {
    new BoolFalse().accept(boolNormalizeVisitor, this) should be(GBool(false))
  }

  val groundNormalizeVisitor: GroundNormalizeVisitor = new GroundNormalizeVisitor() {}

  "A GroundNormalizeVisitor" should "convert a GroundInt to a GInt" in {
    new GroundInt(7).accept(groundNormalizeVisitor, this) should be(GInt(7))
  }

  it should "convert a GroundString to a GString" in {
    val gs = new GroundString("String")

    gs.accept(groundNormalizeVisitor, this) should be(GString("String"))
  }

  it should "convert a GroundUri to a GUri" in {
    val gu = new GroundUri("Uri")

    gu.accept(groundNormalizeVisitor, this) should be(GUri("Uri"))
  }

  // etc, etc....
}
```

File: rholang/src/main/scala/rholang/interpreter/Normalize.scala (hunk 9 above)
Comment (guardbotmk3) [reply rholang/src/main/scala/rholang/interpreter/Normalize.scala]: I'm going to lowercase the filename and leave it for now.

File: rholang/src/main/scala/rholang/interpreter/Normalize.scala [hunk 14]
```diff
@@ -22,18 +22,58 @@ trait GroundNormalizeVisitor extends AbsynGround.Visitor[Ground, Any]
...
 
+trait NameNormalizeVisitor extends Name.Visitor[NameVisitOutputs, NameVisitInputs] {
+  val procVisitor : Proc.Visitor[ProcVisitOutputs, ProcVisitInputs]
+
+  override def visit( n: NameWildcard, input: NameVisitInputs ) = {
+    NameVisitOutputs(ChanVar(WildCard()), input.knownFree.setWildcardUsed())
+  }
+  override def visit( n: NameVar, input: NameVisitInputs ) = {
+    input.env.get(n.var_) match {
+      case Some( (level, NameSort) ) => {
```
[the remark this answers is not in the document]
Comment (guardbotmk3) [reply rholang/src/main/scala/rholang/interpreter/Normalize.scala]: Is it standard style to omit them? I like them.

File: rholang/src/main/scala/rholang/interpreter/Normalize.scala (hunk 11 above)
Comment (guardbotmk3) [reply rholang/src/main/scala/rholang/interpreter/Normalize.scala]: renamed for now.

File: rholang/src/test/scala/rholang/interpreter/NormalizeTest.scala (hunk 10 above)
Comment (guardbotmk3) [reply rholang/src/test/scala/rholang/interpreter/NormalizeTest.scala]: done.

**Comment by guardbotmk3** [comment]
The formatting is fixed. I can split things if you think it's necessary.

File: rholang/src/test/scala/rholang/interpreter/NormalizeTest.scala (hunk 13 above)
Comment (henrytill) [reply rholang/src/test/scala/rholang/interpreter/NormalizeTest.scala]: Custom.

**Pull request merged**